newton/_src/sim/builder.py

```
@@ -1827,6 +1815,18 @@ def add_articulation(
 
+        # Validate that only root bodies (parent == -1) can be kinematic
+        for joint_idx in joints:
+            child = self.joint_child[joint_idx]
+            parent = self.joint_parent[joint_idx]
+            if self.body_flags[child] & BodyFlags.KINEMATIC and parent != -1:
+                child_label = self.body_label[child]
+                parent_label = self.body_label[parent]
+                raise ValueError(
+                    f"Body {child} ('{child_label}') is kinematic but is attached to parent body {parent} "
+                    f"('{parent_label}'). Only root bodies (whose joint parent is the world) can be kinematic."
+                )
```

Ruff (0.15.4)

[warning] Avoid specifying long messages outside the exception class

(TRY003)